src/hotspot/share/oops/method.cpp

```diff
@@ -238,27 +243,73 @@ int Method::fast_exception_handler_bci_for(const methodHandle& mh, Klass* ex_kla
+
     if (beg_bci <= throw_bci && throw_bci < end_bci) {
       // exception handler bci range covers throw_bci => investigate further
+      log_debug(exceptions)("    - entry covers throw point BCI %d", throw_bci);
+
       int handler_bci = table.handler_pc(i);
       int klass_index = table.catch_type_index(i);
       if (klass_index == 0) {
+        if (log_is_enabled(Debug, exceptions)) {
+          log_debug(exceptions)("    - found catch-all handler at BCI: %d", handler_bci);
+        } else if (log_is_enabled(Info, exceptions)) {
+          ResourceMark rm(THREAD);
+          log_info(exceptions)("Found catch-all handler for exception of type \"%s\" in method \"%s\" at BCI: %d",
+                               ex_klass == NULL ? "NULL" : ex_klass->external_name(), mh()->name()->as_C_string(), handler_bci);
+        }
         return handler_bci;
       } else if (ex_klass == NULL) {
+        // Is this even possible?
+        if (log_is_enabled(Debug, exceptions)) {
+          log_debug(exceptions)("    - NULL exception class is implicitly caught by handler at BCI: %d", handler_bci);
+        } else if (log_is_enabled(Info, exceptions)) {
+          ResourceMark rm(THREAD);
+          log_info(exceptions)("NULL exception class is implicitly caught by handler in method \"%s\" at BCI: %d",
+                               mh()->name()->as_C_string(), handler_bci);
+        }
         return handler_bci;
       } else {
+        if (log_is_enabled(Debug, exceptions)) {
+          ResourceMark rm(THREAD);
+          log_debug(exceptions)("    - resolving catch type \"%s\"",
+                               pool->klass_name_at(klass_index)->as_C_string());
+        }
         // we know the exception class => get the constraint class
         // this may require loading of the constraint class; if verification
         // fails or some other exception occurs, return handler_bci
-        Klass* k = pool->klass_at(klass_index, CHECK_(handler_bci));
+        Klass* k = pool->klass_at(klass_index, THREAD);
+        if (HAS_PENDING_EXCEPTION) {
+        if (log_is_enabled(Debug, exceptions)) {
+            ResourceMark rm(THREAD);
+            log_debug(exceptions)("    - exception \"%s\" occurred resolving catch type",
+                                 PENDING_EXCEPTION->klass()->external_name());
+          }
+          return handler_bci;
+        }
         assert(k != NULL, "klass not loaded");
         if (ex_klass->is_subtype_of(k)) {
+          if (log_is_enabled(Debug, exceptions)) {
+            log_debug(exceptions)("    - found matching handler at BCI: %d", handler_bci);
+          } else if (log_is_enabled(Info, exceptions)) {
+            ResourceMark rm(THREAD);
+            log_info(exceptions)("Found matching handler for exception of type \"%s\" in method \"%s\" at BCI: %d",
+                                 ex_klass == NULL ? "NULL" : ex_klass->external_name(), mh()->name()->as_C_string(), handler_bci);
+          }
```

Comment:
There's so much extra code added with this logging.  It's hard to follow as it is. Could you just have the log_info level output in both this and the case above where you find the handler, here and above?  debug level logging will include info level debugging.

Reply:
Logging is supposed to be non-invasive in the source code, which is why it was defined ideally to have one line like log_info(tag)(string)
Here it seems to dominate the code in this method.  Anything to make it shorter and less invasive would be better so that you can see how the code is actually finds the exception handler.  I only made a small request for simplification.  I think the formatting of the log messages wouldn't be bad if the info line is printed rather than the debug line.

[0.208s][debug][exceptions] Looking for catch handler for exception of type "java.lang.Error" in method "main"
[0.208s][debug][exceptions] - checking exception table entry for BCI 0 to 3
[0.208s][debug][exceptions] - entry covers throw point BCI 0
[0.208s][debug][exceptions] - found catch-all handler at BCI: 14

Would be:

[0.208s][debug][exceptions] Looking for catch handler for exception of type "java.lang.Error" in method "main"
[0.208s][debug][exceptions] - checking exception table entry for BCI 0 to 3
[0.208s][debug][exceptions] - entry covers throw point BCI 0
[0.198s][info     ][exceptions] Found catch-all handler for exception of type "java.lang.Error" in method "main" at BCI: 14

I don't think this looks bad at all and worth not having special if statement for debug/info logging.  That's 6 lines saved in this 96 line function.

Reply:
Okay I will redo this.